File: Tests/LedgelingsCoreTests/MeetingsTests.swift
import CoreGraphics
import Testing
@testable import LedgelingsCore

@Suite struct MeetingsTests {
    /// Two creatures of body half-size 10 on the bottom edge, `gap` points apart.
    func pair(gap: CGFloat, segment: (Int, Int) = (0, 0), loop: (Int, Int) = (0, 0),
              canTalk: (Bool, Bool) = (true, true)) -> [Meetings.Party] {
        [Meetings.Party(loop: loop.0, segment: segment.0, position: CGPoint(x: 100, y: 10), halfSize: 10, canTalk: canTalk.0),
         Meetings.Party(loop: loop.1, segment: segment.1, position: CGPoint(x: 120 + gap, y: 10), halfSize: 10, canTalk: canTalk.1)]
    }

    @Test func walkingIntoEachOtherBumpsOnceNotEveryFrame() {
        var meetings = Meetings()
        #expect(meetings.update(pair(gap: 40), at: 0).isEmpty)
        let bumps = meetings.update(pair(gap: 4), at: 1)
        #expect(bumps == [Meetings.Bump(a: 0, b: 1, count: 1, gift: false)])
        #expect(meetings.update(pair(gap: 2), at: 1.1).isEmpty)
        #expect(meetings.update(pair(gap: 0), at: 1.2).isEmpty)
    }

    @Test func onlyCreaturesOnTheSameEdgeOfTheSameMonitorBump() {
        var meetings = Meetings()
        #expect(meetings.update(pair(gap: 0, segment: (0, 1)), at: 1).isEmpty)
        #expect(meetings.update(pair(gap: 0, loop: (0, 1)), at: 2).isEmpty)
    }

    @Test func aSleeperOrJumperDoesNotBump() {
        var meetings = Meetings()
        #expect(meetings.update(pair(gap: 0, canTalk: (true, false)), at: 1).isEmpty)
    }

    @Test func aPairWaitsOutTheCooldownBeforeBumpingAgain() {
        var meetings = Meetings(cooldown: 60)
        #expect(meetings.update(pair(gap: 0), at: 0).count == 1)
        #expect(meetings.update(pair(gap: 100), at: 10).isEmpty)
        #expect(meetings.update(pair(gap: 0), at: 20).isEmpty)          // too soon
        #expect(meetings.update(pair(gap: 100), at: 30).isEmpty)
        #expect(meetings.update(pair(gap: 0), at: 61).count == 1)      // cooldown over
    }

    @Test func theThirdBumpIsAGiftAndTheCountStartsOver() {
        var meetings = Meetings(cooldown: 0, giftEvery: 3)
        var seen: [Meetings.Bump] = []
        for i in 0..<4 {
            let t = Double(i * 10)
            seen += meetings.update(pair(gap: 0), at: t)
            #expect(meetings.update(pair(gap: 100), at: t + 5).isEmpty)
        }
        #expect(seen.map(\.count) == [1, 2, 3, 1])
        #expect(seen.map(\.gift) == [false, false, true, false])
    }

    @Test func threeOnOneEdgeReportEveryTouchingPair() {
        var meetings = Meetings()
        var three = pair(gap: 0)
        three.append(Meetings.Party(loop: 0, segment: 0, position: CGPoint(x: 140, y: 10), halfSize: 10, canTalk: true))
        let bumps = meetings.update(three, at: 1)
        #expect(bumps.map { [$0.a, $0.b] } == [[0, 1], [1, 2]])
    }
}

File: win/LedgelingsCore.Tests/MeetingsTests.cs
namespace Ledgelings.Core.Tests;

public class MeetingsTests
{
    /// <summary>Two creatures of body half-size 10 on the bottom edge, <paramref name="gap"/> points apart.</summary>
    static List<Meetings.Party> Pair(double gap, (int, int)? segment = null, (int, int)? loop = null, (bool, bool)? canTalk = null)
    {
        var seg = segment ?? (0, 0);
        var lp = loop ?? (0, 0);
        var talk = canTalk ?? (true, true);
        return new List<Meetings.Party>
        {
            new(lp.Item1, seg.Item1, new Pt(100, 10), 10, talk.Item1),
            new(lp.Item2, seg.Item2, new Pt(120 + gap, 10), 10, talk.Item2),
        };
    }

    [Fact]
    public void WalkingIntoEachOtherBumpsOnceNotEveryFrame()
    {
        var meetings = new Meetings();
        Assert.Empty(meetings.Update(Pair(40), 0));
        var bumps = meetings.Update(Pair(4), 1);
        Assert.Equal(new[] { new Meetings.Bump(0, 1, 1, false) }, bumps);
        Assert.Empty(meetings.Update(Pair(2), 1.1));
        Assert.Empty(meetings.Update(Pair(0), 1.2));
    }

    [Fact]
    public void OnlyCreaturesOnTheSameEdgeOfTheSameMonitorBump()
    {
        var meetings = new Meetings();
        Assert.Empty(meetings.Update(Pair(0, segment: (0, 1)), 1));
        Assert.Empty(meetings.Update(Pair(0, loop: (0, 1)), 2));
    }

    [Fact]
    public void ASleeperOrJumperDoesNotBump()
    {
        var meetings = new Meetings();
        Assert.Empty(meetings.Update(Pair(0, canTalk: (true, false)), 1));
    }

    [Fact]
    public void APairWaitsOutTheCooldownBeforeBumpingAgain()
    {
        var meetings = new Meetings(cooldown: 60);
        Assert.Single(meetings.Update(Pair(0), 0));
        Assert.Empty(meetings.Update(Pair(100), 10));
        Assert.Empty(meetings.Update(Pair(0), 20));          // too soon
        Assert.Empty(meetings.Update(Pair(100), 30));
        Assert.Single(meetings.Update(Pair(0), 61));         // cooldown over
    }

    [Fact]
    public void TheThirdBumpIsAGiftAndTheCountStartsOver()
    {
        var meetings = new Meetings(cooldown: 0, giftEvery: 3);
        var seen = new List<Meetings.Bump>();
        for (int i = 0; i < 4; i++)
        {
            var t = (double)(i * 10);
            seen.AddRange(meetings.Update(Pair(0), t));
            Assert.Empty(meetings.Update(Pair(100), t + 5));
        }
        Assert.Equal(new[] { 1, 2, 3, 1 }, seen.Select(b => b.Count));
        Assert.Equal(new[] { false, false, true, false }, seen.Select(b => b.Gift));
    }

    [Fact]
    public void ThreeOnOneEdgeReportEveryTouchingPair()
    {
        var meetings = new Meetings();
        var three = Pair(0);
        three.Add(new Meetings.Party(0, 0, new Pt(140, 10), 10, true));
        var bumps = meetings.Update(three, 1);
        Assert.Equal(new[] { new[] { 0, 1 }, new[] { 1, 2 } }, bumps.Select(b => new[] { b.A, b.B }));
    }
}
